test/metabase/query_processor/card_test.clj

```diff
@@ -154,10 +154,50 @@
                   (is (= nil
                          (validate disallowed-type))))))))))
 
-    (testing "Happy path -- API request should succeed if parameter is valid"
+    (testing "Happy path -- API request should succeed if parameter correlates to a template tag by ID"
+      (is (= [1000]
+             (mt/first-row (mt/user-http-request :rasta :post (format "card/%d/query" card-id)
+                                                 {:parameters [{:id    "_DATE_"
+                                                                :type  :date/single
+                                                                :value "2016-01-01"}]})))))
+
+    (testing "Happy path -- API request should succeed if parameter correlates to a template tag by name"
+      (is (= [1000]
+             (mt/first-row (mt/user-http-request :rasta :post (format "card/%d/query" card-id)
+                                                 {:parameters [{:name  "date"
+                                                                :type  :date/single
+                                                                :value "2016-01-01"}]})))))))
+
+(deftest validate-card-parameters-test
+  (mt/with-temp Card [{card-id :id} {:dataset_query (mt/mbql-query checkins {:aggregation [[:count]]})
+                                     :parameters [{:id   "_DATE_"
+                                                   :type :date/single
+                                                   :name "Date"
+                                                   :slug "DATE"}]}]
+    (testing "API request should fail if request parameter does not contain ID"
+      (is (schema= {:message            #"Invalid parameter: Card [\d,]+ does not have a parameter with the ID nil."
```

Comment:
It would be a little nicer if the error message was something along the lines of "invalid parameter: missing `:id`" or something. I wouldn't say it's a dealbreaker since I think that was how it already worked but when I look at this it looks like a little bit of a misleading error message